File: controller.py
from funcs import *
from model import Task


def getDue(t, date):
    if t["due"] != t["date"]:
        return timedeltatostr(inttodate(t.get("due")) - date)
    else:
        return "N/A"


def getAssigned(t, date):
    if t["date"] != datetoint(datetime(2000, 1, 1)):
        return timedeltatostr(inttodate(t.get("date")) - date)
    else:
        return "N/A"


def sortByDue(tasks):
    return sorted(tasks, key=lambda i: i.get("due", float("inf")))


# TODO: I could also make it so that it takes an input, a list of time when i'm busy, or something
# And tell me how many hours i have left to work on something
# TODO: Add percentage, percentage time of due date done, so if half of due date has been exhausted then 50%
def taskDictShorten(t, date):
    taskdict = {
        "name": t["name"],
        "due": getDue(t, date),
        "assigned": getAssigned(t, date),
    }
    if t["desc"] != "":
        taskdict["desc"] = t["desc"]
    taskdict["date"] = t["date"]
    return taskdict


def createTask(task: Task, date: datetime, repeats):
    return {
        "name": task.name(date, repeats),
        "date": datetoint(date),
        "desc": task.description(date),
        "due": datetoint(task.duetime(date)),
        "prio": task.importance(date),
        "ver": task.version(date),
        "num": repeats,  # num repeats
    }


# TODO: i just had the best fucking idea ever, how about instead of parsing into done
# I make it create another json, donetasks.json, with donetasks but in proper format
# so i dont have to kill myself again and again
def getAllUndone():
    pass


# TODO: implement lookahead/foresight, and make things not disappear or whatever
def getToDo(date: datetime, tasks: list, done: list):
    todo = []
    donetasks = {(d["name"], d["date"]) for d in done}
    for i in tasks:
        if inttodate(i["date"]) <= date and ((i["name"], i["date"]) not in donetasks):
            todo.append(i)
    return todo


def getAllTodo(date: datetime):  # TODO: print all even not done
    pass


def getUpcoming(datetime):
    pass


def getAll():
    pass


def getAllDone():
    pass


def getAllDoneToday():  # could make this function be implemented in get all done
    pass


def getTasks(tasks):
    tasklist = list()
    for task in tasks:
        tasklist += evaluate(task)
    return tasklist


def evaluate(task: Task) -> list:
    taskdicts = list()
    check = task.checkstart()
    checkend = task.checkend(check)
    checkrepeats = task.checkrepeats(check, -1)
    repeats = 0
    if task.conditions == []:
        taskdicts.append(createTask(task, check, repeats))
        return taskdicts
    while checkend >= check and checkrepeats != 0:
        if all(fn(check) for fn in task.conditions):
            taskdicts.append(createTask(task, check, repeats))
            checkrepeats = task.checkrepeats(check, checkrepeats)
            repeats += 1
        check += task.checkstep(check)
    return taskdicts

File: src/controller.lua
local funcs = require("src/funcs")

local inttodate = funcs.inttodate
local datetoint = funcs.datetoint
local timedeltatostr = funcs.timedeltatostr

-- ─── Display helpers ──────────────────────────────────────────────────────────

local function getDue(t, date)
	if t.due ~= t.date then
		return timedeltatostr(inttodate(t.due) - date)
	else
		return "N/A"
	end
end

local function getAssigned(t, date)
	local epoch2000 = os.time({ year = 2000, month = 1, day = 1, hour = 0, min = 0, sec = 0 })
	if inttodate(t.date) ~= epoch2000 then
		return timedeltatostr(inttodate(t.date) - date)
	else
		return "N/A"
	end
end

local function getPercentageDone(t, date)
	if t.date ~= t.due then
		return (date - inttodate(t.date)) / (inttodate(t.due) - inttodate(t.date))
	else
		return 0
	end
end

local function sortByDue(tasks)
	local sorted = {}
	for _, v in ipairs(tasks) do
		table.insert(sorted, v)
	end
	table.sort(sorted, function(a, b)
		local ad = a.due or math.huge
		local bd = b.due or math.huge
		return ad < bd
	end)
	return sorted
end

-- TODO: I could also make it so that it takes an input, a list of time when i'm busy, or something
-- And tell me how many hours i have left to work on something
local function taskDictShorten(t, date)
	local taskdict = {
		name = t.name,
		due = getDue(t, date),
		assigned = getAssigned(t, date),
	}
	if t.desc ~= "" then
		taskdict.desc = t.desc
	end
	taskdict["%"] = math.floor(getPercentageDone(t, date) * 100)
	taskdict.date = t.date
	return taskdict
end

local function shortDictPrint(dict)
	for _, t in ipairs(dict) do
		print(string.format("[%3d%%] %-30s assigned: %-20s due: %s", t["%"], t.name, t.assigned, t.due or "N/A"))
	end
end

local function doneTaskPrint(dict, date)
	for _, t in ipairs(dict) do
		print(string.format("%-30s done: %-21s assigned: %-20s", t.name, timedeltatostr(t.done - date), t.date))
	end
end
-- ─── Task creation ────────────────────────────────────────────────────────────

local function createTask(task, date, repeats, hashfunc)
	return {
		name = task.name(date, repeats),
		date = datetoint(date),
		finish = task.finishdelta(date),
		show = task.showdelta(date),
		desc = task.description(date),
		due = datetoint(task.duetime(date)),
		type = task.category(date),
		hash = tostring(hashfunc),
		consec = task.consecutive(date),
		num = repeats,
	}
end

local function getAllUndone(tasks, done)
	done = done or {}
	local donetasks = {}
	for _, d in ipairs(done) do
		donetasks[d.name] = donetasks[d.name] or {}
		donetasks[d.name][d.date] = donetasks[d.name][d.date] or {}
		table.insert(donetasks[d.name][d.date], d.done)
	end

	local undone = {}
	for _, i in ipairs(tasks) do
		if donetasks[i.name] and donetasks[i.name][i.date] then
			local found = false
			for _, value in ipairs(donetasks[i.name][i.date]) do
				if i.finish >= 0 then
					if value >= funcs.inttodate(i.date) and value <= funcs.inttodate(i.due) + i.finish then
						found = true
						break
					end
				else
					if value >= funcs.inttodate(i.date) then
						found = true
						break
					end
				end
			end
			if not found then
				table.insert(undone, i)
			end
		else
			table.insert(undone, i)
		end
	end

	return undone
end

local function getAllDue(date, tasks)
	local due = {}
	for _, value in ipairs(tasks) do
		if inttodate(value.date) - value.show <= date then
			table.insert(due, value)
		end
	end
	return due
end

local function getUnfinished(date, tasks) -- this function for sure works btw i checked
	local unfinished = {}
	for _, value in ipairs(tasks) do
		if inttodate(value.due) + value.finish >= date or (value.finish < 0) then
			table.insert(unfinished, value)
		end
	end
	return unfinished
end

local function getCategory(tasks, Category)
	local catlist = {}
	for _, value in ipairs(tasks) do
		if value.type == Category then
			table.insert(catlist, value)
		end
	end
	return catlist
end

local function getUnconsecutive(tasks)
	local tablex = require("pl.tablex")
	local conlist = tablex.filter(tasks, function(task)
		return task.consec
	end)
	local unconlist = tablex.filter(tasks, function(task)
		return not task.consec
	end)
	for _, task1 in ipairs(conlist) do
		if
			not tablex.find_if(tasks, function(task2)
				return task1.date > task2.date and task1.hash == task2.hash
			end)
		then
			table.insert(unconlist, task1)
		end
	end
	return unconlist
end

local function getToDo(date, tasks, done)
	local result = getUnfinished(date, getAllUndone(getAllDue(date, tasks), done))
	return getUnconsecutive(result)
	-- return getCategory(getUnconsecutive(result), "tasks.test")
	-- return getAllUndone(getUnfinished(date, getAllDue(date, tasks)), done)
end

local function getDoneToday(date, done)
	local donetoday = {}
	for _, value in ipairs(done) do
		if date - 86400 <= value["done"] and date >= value["done"] then
			table.insert(donetoday, value)
		end
	end
	return donetoday
end

local function getUpcoming(date)
	-- pass
end

local function getAll()
	-- pass
end

local function getDoneToFullTasks() -- could make this function be implemented in get all done
	-- pass
end

-- ─── Evaluation ───────────────────────────────────────────────────────────────

local function evaluate(task, date)
	local taskdicts = {}
	local check = task.checkstart(date)
	local checkend = task.checkend(check)
	local checkrepeats = task.checkrepeats(check, -1)
	local repeats = 0
	local hashfunc = {}

	while checkend >= check and checkrepeats ~= 0 do
		local all = true
		for _, fn in ipairs(task.conditions) do
			if not fn(check) then
				all = false
				break
			end
		end
		if all then
			table.insert(taskdicts, createTask(task, check, repeats, hashfunc))
			checkrepeats = task.checkrepeats(check, checkrepeats)
			repeats = repeats + 1
		end
		check = check + task.checkstep(check)
	end

	return taskdicts
end

local function getTasks(tasks, date)
	tasks = tasks or {}
	local tasklist = {}
	for _, task in ipairs(tasks, date) do
		local evaluated = evaluate(task, date)
		for _, t in ipairs(evaluated) do
			table.insert(tasklist, t)
		end
	end
	return tasklist
end

-- ─── Exports ──────────────────────────────────────────────────────────────────

return {
	getDue = getDue,
	getAssigned = getAssigned,
	getPercentageDone = getPercentageDone,
	sortByDue = sortByDue,
	taskDictShorten = taskDictShorten,
	createTask = createTask,
	getToDo = getToDo,
	getTasks = getTasks,
	evaluate = evaluate,
	getAllUndone = getAllUndone,
	getAllTodo = getAllTodo,
	getUpcoming = getUpcoming,
	getAll = getAll,
	-- getAllDone = getAllDone,
	getAllDoneToday = getAllDoneToday,
	getDoneToFullTasks = getDoneToFullTasks,
	getAllDue = getAllDue,
	shortDictPrint = shortDictPrint,
	getUnfinished = getUnfinished,
	getDoneToday = getDoneToday,
	doneTaskPrint = doneTaskPrint,
	getUnconsecutive = getUnconsecutive,
	getCategory = getCategory,
}
